### PR TITLE
#2: Add `sum` function which adds all values in an array

### PR DESCRIPTION
**Pull request by Binney**
Same as #3 except this one's a sum.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -333,6 +333,18 @@ max([2.5, 3.25, -2, 5.75]); // -> 5.75
 
 If the data array is empty or contains a non finite `Number`, the method returns `undefined`.
 
+#### sum
+
+Return the sum of all values in a numeric data array.
+
+Uses the [Kahan summation algorithm](https://en.wikipedia.org/wiki/Kahan_summation_algorithm) to reduce error.
```
Comment (boristane) [inline README.md]: Please provide further details on the algorithm.

File: src/descriptive.js [hunk 2]
```diff
@@ -28,10 +28,31 @@ export function max(arr) {
...
 }
 
+/**
+ * Return the sum of all entries in a numeric array.
+ * @param {Number[]} arr the data array
+ * @returns {Number} the sum of all entries in the array
+ */
+export function sum(arr) {
+  if (!Array.isArray(arr)) return undefined;
+  if (arr.length === 0) return 0;
```
Comment (boristane) [inline src/descriptive.js]: As stated in the doc, this should return undefined.
```suggestion
  if (!Array.isArray(arr) || arr.length === 0) return undefined;
```

File: src/descriptive.js [hunk 3]
```diff
@@ -28,10 +28,31 @@ export function max(arr) {
...
+  // See https://en.wikipedia.org/wiki/Kahan_summation_algorithm
```
Comment (boristane) [inline src/descriptive.js]: Please remove.

File: src/descriptive.js [hunk 4]
```diff
@@ -28,10 +28,31 @@ export function max(arr) {
   return result;
 }
 
+/**
+ * Return the sum of all entries in a numeric array.
```
Comment (boristane) [inline src/descriptive.js]: Please provide further details on the summation algorithm.

File: src/descriptive.js [hunk 5]
```diff
@@ -28,10 +28,31 @@ export function max(arr) {
...
+ */
+export function sum(arr) {
+  if (!Array.isArray(arr)) return undefined;
+  if (arr.length === 0) return 0;
+  // See https://en.wikipedia.org/wiki/Kahan_summation_algorithm
+  let result = 0;
+  let c = 0;
+  for (let i = 0; i < arr.length; i += 1) {
+    if (!Number.isFinite(arr[i])) return undefined;
+    let y = arr[i] - c;
```
Comment (boristane) [inline src/descriptive.js]: Please use `const` where appropriate.
```suggestion
    const y = arr[i] - c;
```

File: src/descriptive.js [hunk 6]
```diff
@@ -28,10 +28,31 @@ export function max(arr) {
...
+    let t = result + y;
```
Comment (boristane) [inline src/descriptive.js]: Please use `const` where appropriate.
```suggestion
    const t = result + y;
```

File: test/descriptive.test.js [hunk 7]
```diff
@@ -30,6 +31,26 @@ describe('Descriptive Statistics', () => {
     testUndefinedWithNullable(max);
   });
 
+  test('Sum', () => {
+    expect(sum([])).toBe(0);
```
Comment (boristane) [inline test/descriptive.test.js]: Should be `undefined`.
```suggestion
    expect(sum([])).toBeUndefined();
```

File: test/descriptive.test.js [hunk 8]
```diff
@@ -30,6 +31,26 @@ describe('Descriptive Statistics', () => {
...
+    expect(sum([3])).toBe(3);
+    expect(sum([1, 2, 3, 4, 5])).toBe(15);
+    expect(sum([0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9,1.0, 1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 1.7]))
+      .toBe(15.3);
+
+    var veryLongArray = [10000, 0];
+    for (let i = 1; i < 501; i++) {
+      veryLongArray[i] = 0.000001;
+    }
+    expect(sum(veryLongArray)).toBe(10000.0005);
```
Comment (boristane) [inline test/descriptive.test.js]: This is not necessary I think. Could you please explain the reasoning behind this ?

File: README.md [hunk 9]
```diff
@@ -333,6 +333,18 @@ max([2.5, 3.25, -2, 5.75]); // -> 5.75
 
 If the data array is empty or contains a non finite `Number`, the method returns `undefined`.
 
+#### sum
```
Comment (boristane) [inline README.md]: Please also update Table of contents.

**Comment by boristane** [comment]
> Same as #3 except this one's a sum.

Please use the standards specified [here](https://blog.github.com/2013-05-14-closing-issues-via-pull-requests/) to close issues with PRs.

File: test/descriptive.test.js (hunk 8 above)
Comment (Binney) [reply test/descriptive.test.js]: Hi! I wanted to ensure the error accumulation was working properly but your example of sum(0.1 through 1.7 step 0.1) demonstrates it well enough compared to the naive algorithm. Have removed.

File: README.md [hunk 10]
```diff
@@ -35,6 +35,7 @@ npm i stat-methods
 3.  [Descriptive statistics](#Descriptive-statistics)
     -   [min](#min)
     -   [max](#max)
+    -   [sum](#sum)
```
Comment (boristane) [inline README.md]: Thanks for this. Would you please also add this to the smaller table of content at below line `306` ? Same for `product` please.

File: src/descriptive.js [hunk 11]
```diff
@@ -29,22 +29,24 @@ export function max(arr) {
 }
 
 /**
- * Return the sum of all entries in a numeric array.
+ * Return the sum of all entries in a numeric array,
+ * using the Kahan summation algorithm.
+ * This compensates for loss of precision when the result is large compared to
+ * the smallest items in the array.
+ * See https://en.wikipedia.org/wiki/Kahan_summation_algorithm
```
Comment (boristane) [inline src/descriptive.js]: Please remove. Not necessary as the name of the algorithm is reference and the doc provides a link to the wikipedia.

File: src/descriptive.js [hunk 12]
```diff
@@ -29,22 +29,24 @@ export function max(arr) {
...
  * @returns {Number} the sum of all entries in the array
  */
 export function sum(arr) {
-  if (!Array.isArray(arr)) return undefined;
-  if (arr.length === 0) return 0;
-  // See https://en.wikipedia.org/wiki/Kahan_summation_algorithm
+  if (!Array.isArray(arr) || arr.length === 0) return undefined;
   let result = 0;
-  let c = 0;
+  let c = 0; // Running compensation of accumulated floating point errors
```
Comment (boristane) [inline src/descriptive.js]: Please remove comments.

File: src/descriptive.js [hunk 13]
```diff
@@ -29,22 +29,24 @@ export function max(arr) {
...
   for (let i = 0; i < arr.length; i += 1) {
     if (!Number.isFinite(arr[i])) return undefined;
-    let y = arr[i] - c;
-    let t = result + y;
-    c = (t - result) - y;
-    result = t;
+    const y = arr[i] - c;
+    const t = result + y; // result is large and y may be small, so least significant bits of y may be lost.
```
Comment (boristane) [inline src/descriptive.js]: Please remove comments.

File: src/descriptive.js [hunk 14]
```diff
@@ -29,22 +29,24 @@ export function max(arr) {
...
+    c = (t - result) - y; // (t - result) cancels the high-order part of y, so we're left with only the least bits
```
Comment (boristane) [inline src/descriptive.js]: Please remove comments.

File: src/descriptive.js [hunk 15]
```diff
@@ -29,22 +29,24 @@ export function max(arr) {
...
+    result = t; // carry over c to the next iteration
```
Comment (boristane) [inline src/descriptive.js]: Please remove comments.

**Pull request closed without merging**